src/nfa/nfa.cc

```diff
@@ -641,34 +641,30 @@ struct StackLevel {
     };
 };
 
-const NumberPredicate<State> Nfa::get_useful_states() const
+std::vector<char> Nfa::get_useful_states() const
```

Comment:
Here char is used to represent bool, this is what the algorithm computes, no need to convert it to something else on return.  We could think about renaming this type somehow, as you mention somewhere else.

Reply:
Yeah. But as mentioned here, we want this function to be in the public interface where one can call this function. They expect a set of states in return. Not some internal structure that has by itself no semantic. Therefore, we should create two functions, one returning the set of states (slower, because it has to compute the set of states first). And second, returning this bit vector (with a new type `BitVector`) which we could use in these optimized algorithms. This second function (implemented above) should be then renamed to something like `BitVector get_useful_states_bitvector()`, explaining that what you get is not a set of states as the name of the function says.